### DISCUSSION TITLE
Why does CONCURRENCY_LEVEL take the value 16?

### DISCUSSION THREAD

**Discussion by hustchaya** [async-profiler/async-profiler#901] [General]
question1 : In the recordSample function, CONCURRENCY_LEVEL serves to limit the number of threads in the sampling stack. Is there any background to setting this data value? Will increasing it lead to high overhead in the target jvm?

question2:  If I remove the lock_index in the recordSample function (the lock is not obtained through the thread ID), and replace _calltrace_buffer in the following code snippet with a temporary CallTraceBuffer variable within the function, will there be any pitfalls in this behavior?
    ASGCT_CallFrame* frames = _calltrace_buffer[lock_index]->_asgct_frames;
    jvmtiFrameInfo* jvmti_frames = _calltrace_buffer[lock_index]->_jvmti_frames;

**Comment by apangin** [comment]
Increasing `CONCURRENCY_LEVEL` will increase profiler memory usage. And yes, this may also increase performance overhead, although this highly depends on the number of cores and the application load.

You can create local call trace buffers, but be aware of that the signal handler runs on an application thread, whose available stack size is unknown. Attempting to allocate too much stack space may lead to undefined behavior, including stack overflow, crash, silent data corruption etc. As a rule, signal handlers should use as little stack as possible.
